vividus-bdd-engine/src/main/java/org/vividus/bdd/transformer/ResolvingSelfReferencesEagerlyTransformer.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2019-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.bdd.transformer;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import javax.inject.Named;
+
+import org.apache.commons.lang3.Validate;
+import org.jbehave.core.model.ExamplesTable.ExamplesTableProperties;
+import org.jbehave.core.model.TableParsers;
+import org.jbehave.core.steps.ParameterControls;
+import org.vividus.bdd.util.ExamplesTableProcessor;
+
+@Named("RESOLVING_SELF_REFERENCES_EAGERLY")
+public class ResolvingSelfReferencesEagerlyTransformer implements ExtendedTableTransformer
+{
+    private final ParameterControls parameterControls;
+    private final Pattern placeholderPattern;
+
+    public ResolvingSelfReferencesEagerlyTransformer(ParameterControls parameterControls)
+    {
+        this.parameterControls = parameterControls;
+        placeholderPattern = Pattern.compile(
+                parameterControls.nameDelimiterLeft() + "(.*?)" + parameterControls.nameDelimiterRight());
+    }
+
+    @Override
+    public String transform(String tableAsString, TableParsers tableParsers, ExamplesTableProperties properties)
+    {
+        List<String> tableAsRows = ExamplesTableProcessor.parseRows(tableAsString);
+        List<String> header = tableParsers.parseRow(tableAsRows.get(0), true, properties);
+        List<List<String>> inputRows = ExamplesTableProcessor.parseDataRows(tableAsRows, tableParsers, properties);
+        List<List<String>> resolvedRows = new SelfReferencesResolver(header).resolveRows(inputRows);
+        return ExamplesTableProcessor.buildExamplesTable(header, resolvedRows, properties, true);
+    }
+
+    private final class SelfReferencesResolver
+    {
+        private final List<String> header;
+
+        SelfReferencesResolver(List<String> header)
+        {
+            this.header = header;
+        }
+
+        List<List<String>> resolveRows(List<List<String>> rows)
+        {
+            return rows.stream().map(this::resolveRow).collect(Collectors.toList());
+        }
+
+        private List<String> resolveRow(List<String> row)
+        {
+            int range = Integer.min(row.size(), header.size());
```

Comment:
no reason to calculate header for each row

Reply:
It is possible to pass table like
```
|x|x|x|
|y|y|
|z|z|z|
```
and catch index out of bounds

Reply:
this.headerSize = header.size()